portfolio/src/main/java/com/google/sps/data/TimeRangeGroupArrayList.java

```diff
@@ -0,0 +1,121 @@
+   * to the list for some reason, this list will potentially need to be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    for (TimeRange currentRange : allTimeRanges) {
+      if (currentRange.overlaps(timeRange)) {
+        throw new IllegalArgumentException(
+            "New time range to add cannot overlap with any existing time ranges");
+      }
+    }
+    allTimeRanges.add(timeRange);
+  }
+
+  /** Returns the array list of all time ranges. */
+  public List<TimeRange> getAllTimeRanges() {
+    sortTimeRanges(comparator, ascending);
+    return allTimeRanges;
+  }
+
+  /**
+   * Checks if a time range exists in the collection. For example, if [3:00 - 4:00] is in the
+   * collection, [3:00 - 3:30] is considered to exist as a time range in the collection.
+   */
+  public boolean hasTimeRange(TimeRange timeRangeToCheck) {
+    for (TimeRange timeRange : allTimeRanges) {
+      if (timeRange.contains(timeRangeToCheck)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Delete a time range from the list. Because the list of all time ranges are always kept to be
+   * pairwise disjoint, the potentially two new time ranges resulted from a deletion will not
+   * overlap with any other existing time ranges.
+   *
+   * <p>For example, if the list contains [3:00 - 4:00] and [5:00 - 6:00], deleting [3:15 - 3:30]
+   * will result in [3 - 3:15] and [3:30 - 4] as new time ranges. Trying to delete [4 - 4:15] will
+   * result in an invalid input exception.
+   *
+   * @return the newly modified collection of time ranges.
+   */
+  public List<TimeRange> deleteTimeRange(TimeRange timeRangeToDelete) {
+    for (TimeRange currentRange : allTimeRanges) {
+      if (currentRange.contains(timeRangeToDelete)) {
+        Instant currentRangeStart = currentRange.start();
+        Instant currentRangeEnd = currentRange.end();
+        Instant toDeleteRangeStart = timeRangeToDelete.start();
+        Instant toDeleteRangeEnd = timeRangeToDelete.end();
+
+        allTimeRanges.remove(currentRange);
+
+        // Construct one or two new time ranges after the deletion.
+        if (currentRangeStart.isBefore(toDeleteRangeStart)) {
+          allTimeRanges.add(TimeRange.fromStartEnd(currentRangeStart, toDeleteRangeStart));
+        }
+
+        if (currentRangeEnd.isAfter(toDeleteRangeEnd)) {
+          allTimeRanges.add(TimeRange.fromStartEnd(toDeleteRangeEnd, currentRangeEnd));
+        }
+        // Collections.sort(allTimeRanges, TimeRange.sortByTimeRangeStartTimeAscending);
+        sortTimeRanges(comparator, ascending);
+        return allTimeRanges;
+      }
+    }
+
+    // If at this point the method hasn't returned yet, it means none of the existing
+    // time ranges actually contains the range to delete.
+    // Throw an invalid input exception here.
+    throw new IllegalArgumentException(
+        "The time range to delete must be contained by an existing time range");
```

Comment:
I'm curious here as well why we can only call this with `TimeRange`s that are completely contained in one of the ones already in the list? If the list contains `[3:00, 4:00]` why not allow the user to delete `[3:30-4:30]` and leave behind just `[3:00, 3:30]`? If the caller is in this situation, what are they supposed to do?

Reply:
The time range group is meant to track free time ranges. So I think conceptually it does not make sense for the user to mark 3:30pm-4:30pm as busy, when the user was free from 3-4. So I think I'm okay with throwing an exception here.

Reply:
Maybe we should rename this to `FreeTimeRangeGroup` then? But I don't see why we wouldn't write it more generally. Also the same situation as above applies here, where the caller doesn't really have any way to know before calling this method whether or not the input will be valid.

Also, I don't see why deleting [3:30, 4:30] is any less valid than deleting [3:30, 4:00] would be? And what if the group contained both [3:00, 4:00] and [4:00-5:00], should it be valid to delete [3:30, 4:30] then?